internal/ethapi/api.go

```diff
@@ -1246,44 +1246,80 @@ func (s *BlockChainAPI) EstimateGas(ctx context.Context, args TransactionArgs, b
 	return DoEstimateGas(ctx, s.b, args, bNrOrHash, s.b.RPCGasCap())
 }
 
-// RPCMarshalHeader converts the given header to the RPC output .
-func RPCMarshalHeader(head *types.Header) map[string]interface{} {
-	result := map[string]interface{}{
-		"number":           (*hexutil.Big)(head.Number),
-		"hash":             head.Hash(),
-		"parentHash":       head.ParentHash,
-		"nonce":            head.Nonce,
-		"mixHash":          head.MixDigest,
-		"sha3Uncles":       head.UncleHash,
-		"logsBloom":        head.Bloom,
-		"stateRoot":        head.Root,
-		"miner":            head.Coinbase,
-		"difficulty":       (*hexutil.Big)(head.Difficulty),
-		"extraData":        hexutil.Bytes(head.Extra),
-		"gasLimit":         hexutil.Uint64(head.GasLimit),
-		"gasUsed":          hexutil.Uint64(head.GasUsed),
-		"timestamp":        hexutil.Uint64(head.Time),
-		"transactionsRoot": head.TxHash,
-		"receiptsRoot":     head.ReceiptHash,
+// RPCHeader represents a serializable header of RPC response
+type RPCHeader struct {
+	Number          *hexutil.Big      `json:"number,omitempty"`
+	Hash            *common.Hash      `json:"hash"`
+	ParentHash      common.Hash       `json:"parentHash"`
+	Nonce           *types.BlockNonce `json:"nonce"`
+	MixDigest       common.Hash       `json:"mixHash"`
+	UncleHash       common.Hash       `json:"sha3Uncles"`
+	Bloom           types.Bloom       `json:"logsBloom"`
+	Root            common.Hash       `json:"stateRoot"`
+	Coinbase        *common.Address   `json:"miner"`
+	Difficulty      *hexutil.Big      `json:"difficulty"`
+	Extra           hexutil.Bytes     `json:"extraData"`
+	GasLimit        hexutil.Uint64    `json:"gasLimit"`
+	GasUsed         hexutil.Uint64    `json:"gasUsed"`
+	Time            hexutil.Uint64    `json:"timestamp"`
+	TxHash          common.Hash       `json:"transactionsRoot"`
+	ReceiptHash     common.Hash       `json:"receiptsRoot"`
+	BaseFee         *hexutil.Big      `json:"baseFeePerGas,omitempty"`
+	WithdrawalsHash *common.Hash      `json:"withdrawalsRoot,omitempty"`
+	TotalDifficulty *hexutil.Big      `json:"totalDifficulty"`
```

Comment:
```suggestion
	TotalDifficulty *hexutil.Big      `json:"totalDifficulty,omitEmpty"`
```

Why not omit empty and get rid of the null values?

Reply:
because the totalDifficulty of `get{Header,Block}ByNumber`'s response is `null`

![Uploading image.png…]()